[Abrir este Notebook en Google Colab](https://colab.research.google.com/github/heldigard/Electiva-Analitica-IA/blob/main/Clase1-Electiva-Analitica-IA.ipynb)

# Clase 1: Analítica de Datos usando IA - Exploración Inicial

**Objetivo:** En esta sesión, aprenderemos a utilizar un asistente de Inteligencia Artificial (Gemini en Colab) como nuestro copiloto para realizar los primeros pasos de cualquier proyecto de datos: la carga, exploración y limpieza inicial.

**Caso de Estudio:** Análisis de Siniestros Viales en Bogotá.
**Dataset:** `siniestros_viales_consolidados_bogota_dc.xlsx`

**Pagina Descarga:** https://www.datos.gov.co/dataset/Siniestros-Viales-Consolidados-Bogot-D-C-/9vzw-xjch/about_data

**Link Descarga:** https://datosabiertos.bogota.gov.co/dataset/0b070626-fe5a-42dc-ae99-92601da166d9/resource/9e3add8e-3f46-42ce-9c0f-419964f6f598/download/siniestros_viales_consolidados_bogota_dc.xlsx

**Metodología:** No escribiremos código desde cero. Nuestra tarea será formular las preguntas correctas (prompts) al asistente de IA para que genere el código por nosotros. Luego, interpretaremos los resultados.

**Paso 0: Activar el asistente de IA**
Asegúrate de tener visible el panel de "Asistente de IA" en Colab. Si no lo ves, ve al menú "Herramientas" -> "Configuración" -> "Asistente de IA" y actívalo.

### **Paso 1: Entender nuestro archivo de Excel**

Nuestro archivo (`siniestros_viales_consolidados_bogota_dc.xlsx`) es un Excel, que puede contener múltiples hojas. Antes de cargar los datos, necesitamos saber qué hojas contiene para decidir cuál es la relevante para nuestro análisis.

**Instrucción:** Copia y pega el siguiente prompt en la ventana del asistente de IA.

Prompt para el Asistente de IA:
> Hola. Estoy trabajando en Google Colab y he subido un archivo de Excel llamado 'siniestros_viales_consolidados_bogota_dc.xlsx'. Necesito comenzar a analizarlo. Antes que nada, genera el código de Python usando la biblioteca pandas para que me muestres una lista con los nombres de todas las hojas que contiene este archivo de Excel.



In [ ]:
import pandas as pd

# Nombre del archivo de Excel
file_path = 'siniestros_viales_consolidados_bogota_dc.xlsx'

# Cargar el archivo de Excel
xls = pd.ExcelFile(file_path)

# Obtener y mostrar los nombres de las hojas
sheet_names = xls.sheet_names
print("Las hojas en el archivo son:")
print(sheet_names)

Las hojas en el archivo son:
['SINIESTROS', 'ACTOR_VIAL', 'VEHICULOS', 'HIPOTESIS', 'DICCIONARIO']


**Interpretación:**
Al ejecutar el código, vemos la lista real de las hojas. Basado en los nombres, podemos inferir lo siguiente:
*   **'SINIESTROS'**: Esta parece ser la hoja principal que contiene la información de cada evento o siniestro. Es nuestro punto de partida.
*   **'ACTOR_VIAL'**, **'VEHICULOS'**, **'HIPOTESIS'**: Estas hojas probablemente contienen datos detallados que se pueden vincular al siniestro principal (por ejemplo, información de cada persona involucrada, de cada vehículo, o las posibles causas). Las usaremos más adelante.
*   **'DICCIONARIO'**: Muy útil. Seguramente explica qué significa cada código o variable en las otras hojas.

Comenzaremos nuestro análisis cargando la hoja **'SINIESTROS'**.

### **Paso 2: Cargar la hoja de datos principal**

Ahora le pediremos a nuestro asistente que cargue específicamente la hoja 'SINIESTROS' en una estructura de datos que podamos manipular, llamada DataFrame.

**Instrucción:** Usa el siguiente prompt con el asistente.

Prompt para el Asistente de IA:
> Gracias. La hoja que me interesa para empezar es 'SINIESTROS'. Ahora, genera el código para cargar únicamente la hoja 'SINIESTROS' del archivo 'siniestros_viales_consolidados_bogota_dc.xlsx' en un DataFrame de pandas que se llame df_siniestros. Al final, confirma la carga imprimiendo un mensaje que diga "DataFrame df_siniestros cargado exitosamente".

In [ ]:
import pandas as pd

# Nombre del archivo y de la hoja
file_path = 'siniestros_viales_consolidados_bogota_dc.xlsx'
sheet_name = 'SINIESTROS'

# Cargar la hoja 'SINIESTROS' en un DataFrame
df_siniestros = pd.read_excel(file_path, sheet_name=sheet_name)

print("DataFrame df_siniestros cargado exitosamente.")

DataFrame df_siniestros cargado exitosamente.


### **Paso 3: El primer vistazo a nuestros datos**

La carga fue exitosa, pero no hemos visto nada aún. Un primer paso fundamental es "ver" una muestra de los datos para entender qué tipo de columnas y valores contiene. Le pediremos al asistente que nos muestre las primeras filas.

**Instrucción:** Pídele al asistente que te muestre cómo son los datos.

Prompt para el Asistente de IA:
> Perfecto. Ahora quiero tener un primer vistazo de los datos que acabamos de cargar en el DataFrame df_siniestros. Genera el código para mostrar las primeras 10 filas de este DataFrame.

In [ ]:
# Mostrar las primeras 10 filas del DataFrame
df_siniestros.head(10)

CODIGO_ACCIDENTE       FECHA      HORA  GRAVEDAD  CLASE  CHOQUE  \
0           4401438  01/01/2015  01:05:00         2      2     NaN   
1           4401449  01/01/2015  05:50:00         2      3     NaN   
2           4401430  01/01/2015  07:15:00         2      3     NaN   
3           4401453  01/01/2015  09:30:00         3      1     1.0   
4           4401423  01/01/2015  09:45:00         2      1     1.0   
5           4401437  01/01/2015  12:50:00         3      1     1.0   
6           4401420  01/01/2015  13:00:00         3      1     1.0   
7           4401425  01/01/2015  13:25:00         3      1     1.0   
8           4401447  01/01/2015  13:30:00         3      1     1.0   
9           4401443  01/01/2015  14:30:00         3      1     1.0   

   OBJETO_FIJO                     DIRECCION  CODIGO_LOCALIDAD  DISENO_LUGAR  
0          NaN               KR 64A-CL 2C 02                16             2  
1          NaN  AV AVENIDA DEL SUR-KR 65A 41                 7             1  
2          NaN             KR 19D-CL 62 S 02                19             1  
3          NaN             KR 79-CL 42F S 71                 8             1  
4          NaN               CL 66A-KR 76 02                10             2  
5          NaN              DG 77A-AK 18D 02                19             1  
6          NaN              KR 17A-CL 63B 02                19             1  
7          NaN             CL 143A-KR 142 02                11             1  
8          NaN               CL 72-KR 91A 45                10             1  
9          NaN               KR 100-CL 23 02                 9             1

### **Paso 4: Resumen técnico del DataFrame**

Ver las primeras filas nos da una idea visual, pero para un análisis riguroso necesitamos un resumen técnico. Con este resumen podremos responder preguntas clave como:

1.  ¿Cuántos registros (filas) y variables (columnas) tenemos en total?
2.  ¿Qué nombre y tipo de dato tiene cada columna (texto, número, fecha)?
3.  **Crucial:** ¿Cuántos datos nos faltan (valores nulos o vacíos) en cada columna?

**Instrucción:** Ahora, pide al asistente un resumen completo de la estructura de nuestros datos.

In [ ]:
# Obtener información del DataFrame
df_siniestros.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196152 entries, 0 to 196151
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   CODIGO_ACCIDENTE  196152 non-null  int64  
 1   FECHA             196152 non-null  object 
 2   HORA              196152 non-null  object 
 3   GRAVEDAD          196152 non-null  int64  
 4   CLASE             196152 non-null  int64  
 5   CHOQUE            167910 non-null  float64
 6   OBJETO_FIJO       6689 non-null    float64
 7   DIRECCION         196152 non-null  object 
 8   CODIGO_LOCALIDAD  196152 non-null  int64  
 9   DISENO_LUGAR      196152 non-null  int64  
dtypes: float64(2), int64(5), object(3)
memory usage: 15.0+ MB


**Análisis de los resultados (¡La parte humana!):**
El resultado de `.info()` es la radiografía de nuestro dataset. Analicemos lo que nos muestra con los datos reales:

*   **Total de Registros:** La primera línea nos dice que tenemos **196,152** siniestros registrados. ¡Es un conjunto de datos considerable!

*   **Tipos de Datos (Dtype):**
    *   Tal como sospechábamos, las columnas `FECHA` y `HORA` son de tipo **`object`** (texto). Esto es un problema prioritario. No podemos analizar tendencias temporales si el programa no entiende que son fechas. Nuestra primera gran tarea será corregir esto.
    *   Columnas clave como `GRAVEDAD` y `CODIGO_LOCALIDAD` son `int64` (números), lo cual está bien. Recordaremos que son códigos categóricos.

*   **Valores Nulos (Non-Null Count):** Aquí está el hallazgo más importante.
    *   **Columnas Completas:** ¡Buenas noticias! Columnas vitales como `FECHA`, `HORA`, `GRAVEDAD`, `DIRECCION` y `CODIGO_LOCALIDAD` están 100% completas (196,152 valores no nulos).
    *   **Columna `CHOQUE`:** Tiene **167,910** valores de un total de 196,152. Faltan casi 30,000 datos. Esto probablemente significa que esta columna solo se rellena cuando el siniestro es, efectivamente, un choque entre vehículos. No es necesariamente un error, pero debemos manejar esos vacíos (NaNs).
    *   **Columna `OBJETO_FIJO`:** Este es un caso extremo. Solo tiene **6,689** valores. Más del 96% de los datos en esta columna están vacíos. Una columna tan vacía aporta muy poca información al análisis general y puede causar problemas en pasos futuros. Es una candidata muy fuerte para ser eliminada.

**Plan de Acción:**
Basado en este diagnóstico, nuestras próximas tareas de limpieza y preparación serán:
1.  Eliminar la columna `OBJETO_FIJO` por tener un exceso de valores faltantes.
2.  Unificar las columnas `FECHA` y `HORA` en una sola columna con el formato correcto de fecha y hora (`datetime`).
3.  Investigar y decidir una estrategia para los valores faltantes en la columna `CHOQUE`.

Antes de empezar a modificar el DataFrame, obtengamos un resumen final de las columnas que sí son numéricas para entender sus rangos y distribución.

### **Paso 5: Resumen estadístico**

Ahora, para las columnas que pandas identificó como numéricas, queremos entender sus estadísticas básicas (media, mínimo, máximo, etc.). Esto nos puede ayudar a detectar valores extraños.

**Instrucción:** Pide un resumen estadístico.

Prompt para el Asistente de IA:
> Ahora quiero un resumen estadístico únicamente de las columnas numéricas en mi DataFrame df_siniestros para entender sus distribuciones. Muéstrame la media, la desviación estándar, y los valores mínimo y máximo de cada una.

In [ ]:
# Obtener resumen estadístico de las columnas numéricas
df_siniestros.describe()

CODIGO_ACCIDENTE       GRAVEDAD          CLASE         CHOQUE  \
count      1.961520e+05  196152.000000  196152.000000  167910.000000   
mean       6.997374e+06       2.636873       1.221955       1.127717   
std        2.976910e+06       0.511668       0.653418       0.610152   
min        4.401419e+06       1.000000       1.000000       1.000000   
25%        4.451508e+06       2.000000       1.000000       1.000000   
50%        4.501210e+06       3.000000       1.000000       1.000000   
75%        1.048079e+07       3.000000       1.000000       1.000000   
max        1.053701e+07       3.000000       7.000000       5.000000   

       OBJETO_FIJO  CODIGO_LOCALIDAD   DISENO_LUGAR  
count  6689.000000     196152.000000  196152.000000  
mean      5.318583          9.216424       1.420312  
std       3.756058          5.015592       1.284176  
min       1.000000          1.000000       1.000000  
25%       2.000000          6.000000       1.000000  
50%       5.000000          9.000000       1.000000  
75%      10.000000         12.000000       1.000000  
max      16.000000         20.000000      13.000000

**Interpretación Crítica (¡La Lección Clave!):**
La tabla que genera el asistente está llena de números (media, desviación estándar, etc.), pero ¿tienen sentido?
*   Columnas como `GRAVEDAD`, `CLASE` o `CODIGO_LOCALIDAD` son **códigos categóricos**, no cantidades. ¿Qué significa la "localidad promedio"? ¡Nada! Calcular estadísticas sobre ellas es incorrecto.
*   `CODIGO_ACCIDENTE` es un **identificador único**. Sus estadísticas son irrelevantes.
*   **Conclusión:** Hemos aprendido que no debemos confiar ciegamente en los resultados automáticos. El contexto nos dice que la mayoría de estas estadísticas no son útiles.

### **Paso 6: Interpretando los Códigos con la Hoja 'DICCIONARIO'**

Acabamos de ver el resumen estadístico (`.describe()`) y notamos que muchas columnas como `GRAVEDAD`, `CLASE` y `CHOQUE` son números. Pero, ¿qué significan realmente estos números? ¿El valor `3` en `GRAVEDAD` es más grave que `1`? No necesariamente, son **códigos**.

Para descifrarlos, usaremos la hoja `DICCIONARIO` que descubrimos al principio. Nuestra primera tarea será cargar esta hoja en un nuevo DataFrame para poder consultarla y usarla como nuestra "piedra de Rosetta" para entender el dataset.

**Instrucción:** Pídele al asistente que cargue la hoja `DICCIONARIO`.

Prompt para el Asistente de IA:
> Ya tengo el DataFrame df_siniestros. Ahora necesito cargar otra hoja del mismo archivo de Excel, 'siniestros_viales_consolidados_bogota_dc.xlsx'. Por favor, genera el código para cargar la hoja llamada 'DICCIONARIO' en un nuevo DataFrame que se llame df_diccionario. Luego, muéstrame las primeras 15 filas de este nuevo DataFrame.

In [ ]:
import pandas as pd

# Cargar la hoja 'DICCIONARIO'
df_diccionario = pd.read_excel('siniestros_viales_consolidados_bogota_dc.xlsx', sheet_name='DICCIONARIO')

# Mostrar las primeras 15 filas para ver su contenido
df_diccionario.head(15)

HOJA        CAMPO  CODIGO        DESCRIPCION
0   SINIESTROS     GRAVEDAD       1        Con Muertos
1   SINIESTROS     GRAVEDAD       2        Con Heridos
2   SINIESTROS     GRAVEDAD       3         Solo Daños
3   SINIESTROS        CLASE       7         Autolesion
4   SINIESTROS        CLASE       1             Choque
5   SINIESTROS        CLASE       2          Atropello
6   SINIESTROS        CLASE       4  Caida de ocupante
7   SINIESTROS        CLASE       3        Volcamiento
8   SINIESTROS        CLASE       6               Otro
9   SINIESTROS        CLASE       5           Incendio
10  SINIESTROS       CHOQUE       1           Vehiculo
11  SINIESTROS       CHOQUE       2               Tren
12  SINIESTROS       CHOQUE       3         Semoviente
13  SINIESTROS       CHOQUE       4        Objeto Fijo
14  SINIESTROS  OBJETO_FIJO       1               Muro

**Análisis del Diccionario:**
¡Perfecto! Ahora tenemos una "ayuda" para nuestro análisis. Vemos que tiene 4 columnas: `HOJA`, `CAMPO`, `CODIGO` y `DESCRIPCION`.

Ahora podemos usar este `df_diccionario` para entender nuestras columnas. Por ejemplo, ¿qué significan los valores de la columna `GRAVEDAD` en nuestro `df_siniestros`? Podemos "filtrar" el diccionario para que nos lo diga.

**Instrucción:** Pídele al asistente que te muestre los significados para la columna `GRAVEDAD`.

Prompt para el Asistente de IA:
> Usando el DataFrame df_diccionario, genera el código de Python para mostrarme únicamente las filas donde el valor de la columna CAMPO sea exactamente igual a 'GRAVEDAD'.

In [ ]:
# Filtrar el diccionario para ver los valores de GRAVEDAD
descripcion_gravedad = df_diccionario[df_diccionario['CAMPO'] == 'GRAVEDAD']

print(descripcion_gravedad)

         HOJA     CAMPO  CODIGO  DESCRIPCION
0  SINIESTROS  GRAVEDAD       1  Con Muertos
1  SINIESTROS  GRAVEDAD       2  Con Heridos
2  SINIESTROS  GRAVEDAD       3   Solo Daños


**Interpretación del Código de Gravedad:**
¡Ahora sí! Al ejecutar el código anterior, vemos claramente:
*   **Código 1:** Con Muertos
*   **Código 2:** Con Heridos
*   **Código 3:** Solo Daños

Ahora, cuando veamos un `3` en la columna `GRAVEDAD` de `df_siniestros`, ya no es un número abstracto, sabemos que se refiere a un accidente solo con daños materiales.

**Ejercicio:** Repite el proceso anterior para el campo **`CLASE`**. Modifica el código de la celda anterior para filtrar por `'CLASE'` y entender qué tipos de siniestro existen.

### **Paso 7: La Gran Limpieza (Ahora con Criterio)**

Ahora que entendemos mejor nuestros datos y hemos usado el diccionario para descifrar los códigos, podemos ejecutar nuestro plan de limpieza con total confianza. Recordemos las decisiones basadas en nuestro análisis:

1.  **Eliminar `OBJETO_FIJO`**: Confirmamos que tiene más del 96% de datos faltantes. Una columna tan vacía no aporta información fiable y puede interferir con futuros análisis.
2.  **Eliminar `CODIGO_ACCIDENTE`**: Es un identificador único (como el DNI de una persona), no una variable que describa una característica del siniestro. No sirve para encontrar patrones.
3.  **Eliminar `CODIGO_LOCALIDAD`**: Ya vimos en el Paso 5 que tratar este código como un número para calcular promedios es metodológicamente incorrecto. Lo eliminamos para evitar análisis erróneos. En clases futuras, podríamos aprender a usarlo como una categoría, pero por ahora, está fuera de nuestro alcance analítico.

Vamos a pedirle al asistente que ejecute estas tres acciones en un solo paso.

**Instrucción:** Pídele al asistente que elimine las columnas de una vez.

Prompt para el Asistente de IA:
> Necesito limpiar mi DataFrame `df_siniestros` basándome en el análisis que hemos hecho. Por favor, genera el código para eliminar permanentemente las siguientes tres columnas: 'OBJETO_FIJO', 'CODIGO_ACCIDENTE' y 'CODIGO_LOCALIDAD'. Para confirmar que la operación fue exitosa, después de eliminarlas, imprime un mensaje y la lista de las columnas que quedaron en el DataFrame.

In [ ]:
# Lista de columnas a eliminar
columnas_a_eliminar = ['OBJETO_FIJO', 'CODIGO_ACCIDENTE', 'CODIGO_LOCALIDAD']

# Eliminar las columnas del DataFrame de forma permanente
df_siniestros.drop(columns=columnas_a_eliminar, inplace=True)

# Confirmar la operación
print(f"Se eliminaron las columnas: {columnas_a_eliminar}")
print("Las columnas restantes en el DataFrame son:")
print(df_siniestros.columns)

Se eliminaron las columnas: ['OBJETO_FIJO', 'CODIGO_ACCIDENTE', 'CODIGO_LOCALIDAD']
Las columnas restantes en el DataFrame son:
Index(['FECHA', 'HORA', 'GRAVEDAD', 'CLASE', 'CHOQUE', 'DIRECCION',
       'DISENO_LUGAR'],
      dtype='object')


### **Paso 8: Ingeniería de Características (La Fusión de Fecha y Hora)**

Nuestro DataFrame está limpio, pero todavía tenemos un problema técnico que nos impide hacer análisis avanzados: las columnas `FECHA` y `HORA` son de tipo `object` (texto). Para un computador, "01/01/2015" es solo una cadena de caracteres, no un punto en el tiempo.

Nuestra tarea de preparación más importante es crear una columna de tiempo que el sistema entienda. Vamos a combinar `FECHA` y `HORA` en una sola columna de tipo `datetime`, el formato universal de Python para trabajar con fechas y horas. Este proceso se llama **Ingeniería de Características**: crear nuevas variables más útiles a partir de las existentes.

**Instrucción:** Pídele al asistente que combine, convierta y limpie las columnas de tiempo.

Prompt para el Asistente de IA:
> Es hora de trabajar con las fechas. Necesito hacer una operación de ingeniería de características en mi DataFrame `df_siniestros`. Por favor, genera el código para realizar los siguientes tres pasos en orden:
> 1. Crea una nueva columna llamada `FECHA_HORA` combinando los valores de las columnas `FECHA` y `HORA`.
> 2. Asegúrate de que esta nueva columna `FECHA_HORA` sea de tipo `datetime`, que es el formato correcto para fechas y horas.
> 3. Una vez creada y convertida la nueva columna, elimina las columnas originales `FECHA` y `HORA` que ya no son necesarias.
> Finalmente, muestra las primeras 5 filas del DataFrame para ver el resultado.

In [ ]:
# 1. Combinar FECHA y HORA y 2. Convertir a formato datetime
# Se le pasa el formato esperado de la fecha para asegurar la conversión correcta
df_siniestros['FECHA_HORA'] = pd.to_datetime(df_siniestros['FECHA'] + ' ' + df_siniestros['HORA'], format='%d/%m/%Y %H:%M:%S')

# 3. Eliminar las columnas originales
df_siniestros.drop(columns=['FECHA', 'HORA'], inplace=True)

# Confirmar la operación mostrando el resultado
print("Columna 'FECHA_HORA' creada y columnas originales eliminadas exitosamente.")
df_siniestros.head()

Columna 'FECHA_HORA' creada y columnas originales eliminadas exitosamente.


GRAVEDAD  CLASE  CHOQUE                     DIRECCION  DISENO_LUGAR  \
0         2      2     NaN               KR 64A-CL 2C 02             2   
1         2      3     NaN  AV AVENIDA DEL SUR-KR 65A 41             1   
2         2      3     NaN             KR 19D-CL 62 S 02             1   
3         3      1     1.0             KR 79-CL 42F S 71             1   
4         2      1     1.0               CL 66A-KR 76 02             2   

           FECHA_HORA  
0 2015-01-01 01:05:00  
1 2015-01-01 05:50:00  
2 2015-01-01 07:15:00  
3 2015-01-01 09:30:00  
4 2015-01-01 09:45:00

### **Paso 9: Investigando Anomalías de Tipos de Dato (el caso de `CHOQUE`)**

Acabamos de hacer la verificación con `.head()` y, si somos observadores, notaremos algo extraño. La columna `CHOQUE` contiene valores como `1.0` y `NaN`. El `.info()` del principio nos dijo que es de tipo `float64`.

Esto debería levantar una bandera roja: ¿Por qué una columna de códigos categóricos es un número decimal (`float`) y no un entero (`int`)?

La razón más común en `pandas` es la presencia de valores nulos (`NaN`, Not a Number). `NaN` es técnicamente un tipo `float`, y su presencia "contamina" u obliga a toda la columna a convertirse en `float` para poder albergarlo.

Vamos a confirmar esta hipótesis y entender mejor la columna.

**Instrucción:** Pídele al asistente que cuente los valores únicos en la columna `CHOQUE`, incluyendo los nulos.

Prompt para el Asistente de IA:
> He notado que la columna `CHOQUE` en `df_siniestros` es de tipo float, lo cual es extraño para un código. Sospecho que es por los valores nulos.
> Por favor, genera el código para contar cuántas veces aparece cada valor único en la columna `CHOQUE`. Asegúrate de que el conteo también incluya los valores nulos (`NaN`) para confirmar mi hipótesis.

In [ ]:
# Contar los valores únicos en la columna 'CHOQUE', incluyendo los NaN
# El argumento dropna=False es crucial para que los nulos sean contados.
conteo_choque = df_siniestros['CHOQUE'].value_counts(dropna=False)

print("Distribución de valores en la columna 'CHOQUE':")
print(conteo_choque)

Distribución de valores en la columna 'CHOQUE':
CHOQUE
1.0    160808
NaN     28242
4.0      6689
5.0       290
3.0        95
2.0        28
Name: count, dtype: int64


**Interpretación y Decisión:**

¡Nuestra hipótesis era correcta! La salida del código nos muestra claramente que hay **28,242** valores nulos (`NaN`). Su presencia es lo que obliga a toda la columna `CHOQUE` a ser de tipo `float`.

Además, este análisis nos da contexto: la gran mayoría de los choques son contra otro vehículo (código 1.0, con 160,808 casos). Este tipo de investigación rápida nos ayuda a entender la naturaleza de nuestros datos. Por ahora, dejaremos esta columna como está, entendiendo que los `NaN` simplemente significan "No aplica" (el siniestro no fue un choque).

### **Paso 10: De Códigos a Categorías: Transformando la Columna `CHOQUE`**

Hemos confirmado que la columna `CHOQUE` es de tipo `float` debido a los valores nulos (`NaN`). Sin embargo, los números como `1.0` o `4.0` son solo códigos. Dejarlos así es un riesgo, ya que un futuro análisis podría tratarlos erróneamente como cantidades numéricas.

La solución profesional es doble:
1.  **Hacerla legible:** Usaremos nuestro `df_diccionario` para reemplazar los códigos numéricos por su descripción en texto (ej. `1.0` se convertirá en 'Vehiculo').
2.  **Asignar el tipo correcto:** Una vez que tengamos el texto, la convertiremos al tipo de dato `category`. Este es el tipo de dato ideal en `pandas` para columnas que tienen un número limitado de opciones posibles. Es más eficiente en memoria que el texto (`object`) y representa correctamente la naturaleza de nuestros datos.

**Instrucción:** Pídele al asistente que mapee los códigos de `CHOQUE` a sus descripciones y cambie el tipo de dato.

Prompt para el Asistente de IA:
> Quiero hacer mi columna `CHOQUE` más legible y asignarle el tipo de dato correcto. Por favor, genera un código que realice los siguientes pasos en secuencia:
> 1. Primero, filtra el `df_diccionario` para obtener solo las filas correspondientes al campo 'CHOQUE'.
> 2. Con ese resultado, crea un diccionario de Python que mapee los `CODIGO` a su `DESCRIPCION`.
> 3. Usa este diccionario y el método `.map()` para reemplazar los valores numéricos en la columna `df_siniestros['CHOQUE']`.
> 4. Convierte la columna `CHOQUE` ya mapeada al tipo de dato 'category'.
> 5. Finalmente, para verificar, muéstrame el conteo de valores de la nueva columna `CHOQUE` y luego el resumen `.info()` del DataFrame para confirmar el cambio de tipo de dato.

In [ ]:
# 1. Filtrar el diccionario para obtener los códigos de 'CHOQUE'
filtro_choque = df_diccionario[df_diccionario['CAMPO'] == 'CHOQUE']

# 2. Crear un diccionario de mapeo {codigo: descripcion}
# Usamos un método eficiente para crear el diccionario desde las columnas
mapa_choque = pd.Series(filtro_choque.DESCRIPCION.values, index=filtro_choque.CODIGO).to_dict()

print("El diccionario de mapeo creado es:")
print(mapa_choque)

# 3. Aplicar el mapeo a la columna 'CHOQUE'. Los NaN no se verán afectados.
df_siniestros['CHOQUE'] = df_siniestros['CHOQUE'].map(mapa_choque)

# 4. Convertir la columna al tipo 'category'
df_siniestros['CHOQUE'] = df_siniestros['CHOQUE'].astype('category')

# 5. Verificar los resultados
print("\nConteo de valores en la columna 'CHOQUE' transformada:")
print(df_siniestros['CHOQUE'].value_counts(dropna=False))

print("\nVerificación del tipo de dato final:")
df_siniestros.info()

El diccionario de mapeo creado es:
{1: 'Vehiculo', 2: 'Tren', 3: 'Semoviente', 4: 'Objeto Fijo'}

Conteo de valores en la columna 'CHOQUE' transformada:
CHOQUE
Vehiculo       160808
NaN             28532
Objeto Fijo      6689
Semoviente         95
Tren               28
Name: count, dtype: int64

Verificación del tipo de dato final:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196152 entries, 0 to 196151
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   GRAVEDAD      196152 non-null  int64         
 1   CLASE         196152 non-null  int64         
 2   CHOQUE        167620 non-null  category      
 3   DISENO_LUGAR  196152 non-null  int64         
 4   FECHA_HORA    196152 non-null  datetime64[ns]
dtypes: category(1), datetime64[ns](1), int64(3)
memory usage: 6.2 MB


### **Paso 11: Manejo de Columnas con Demasiada Información (el caso de `DIRECCION`)**

Ahora miremos la columna `DIRECCION`. Aunque parece muy informativa, para un análisis de patrones a alto nivel, presenta un problema: tiene una **alta cardinalidad**. Esto significa que tiene una cantidad enorme de valores únicos (casi uno por cada siniestro).

Si intentáramos agrupar los datos por dirección para ver "en qué dirección hay más accidentes", los resultados no serían útiles, ya que casi cada accidente tendría su propio grupo. Para nuestro análisis estadístico actual, esta columna es más ruido que señal.

**Nota Pedagógica:** ¡Esta columna sería oro puro si quisiéramos hacer un análisis geoespacial y pintar los accidentes en un mapa! Pero para descubrir patrones estadísticos generales y agregados, es mejor quitarla.

**Instrucción:** Pídele al asistente que elimine la columna `DIRECCION`.

Prompt para el Asistente de IA:
> La columna `DIRECCION` tiene demasiados valores únicos, lo que se conoce como alta cardinalidad, y no será útil para el tipo de análisis agregado que estoy planeando. Por favor, genera el código para eliminar permanentemente la columna `DIRECCION` de mi DataFrame `df_siniestros`. Luego, para confirmar, imprime la lista de columnas que quedan.

In [ ]:
# Eliminar la columna 'DIRECCION' debido a su alta cardinalidad
df_siniestros.drop(columns=['DIRECCION'], inplace=True)

# Confirmar que la columna ha sido eliminada
print("La columna 'DIRECCION' ha sido eliminada exitosamente.")
print("Las columnas finales de nuestro DataFrame son:")
print(df_siniestros.columns)

La columna 'DIRECCION' ha sido eliminada exitosamente.
Las columnas finales de nuestro DataFrame son:
Index(['GRAVEDAD', 'CLASE', 'CHOQUE', 'DISENO_LUGAR', 'FECHA_HORA'], dtype='object')


### **Paso 12: La Verificación Final**

Muy bien. Hemos cargado, diagnosticado, interpretado, descifrado códigos y hemos tomado decisiones de limpieza basadas en evidencia, eliminando columnas por exceso de nulos, por ser identificadores, por ser categorías mal representadas y por alta cardinalidad.

El paso final de este riguroso proceso es, y siempre debe ser, **verificar el estado final de nuestro producto.** Vamos a pedirle al asistente una última vez el resumen técnico para tener la "foto final" de nuestro dataset preparado.

**Instrucción:** Pide la verificación final.

Prompt para el Asistente de IA:
> Hemos completado toda la limpieza. Para la verificación final, por favor, genera el código para mostrar de nuevo el resumen técnico (.info()) de mi DataFrame `df_siniestros`. Quiero ver la estructura definitiva con la que trabajaremos.

In [ ]:
# Obtener el resumen técnico final para verificar todos los cambios.
print("Resumen técnico final del DataFrame df_siniestros:")
df_siniestros.info()

Resumen técnico final del DataFrame df_siniestros:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196152 entries, 0 to 196151
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   GRAVEDAD      196152 non-null  int64         
 1   CLASE         196152 non-null  int64         
 2   CHOQUE        167620 non-null  category      
 3   DISENO_LUGAR  196152 non-null  int64         
 4   FECHA_HORA    196152 non-null  datetime64[ns]
dtypes: category(1), datetime64[ns](1), int64(3)
memory usage: 6.2 MB


### **Paso 12: Conclusión y Próximos Pasos**

¡Excelente trabajo! Hemos llegado al final de nuestra primera sesión, y el resultado es un DataFrame completamente transformado. Comparemos nuestra "foto final" (`.info()`) con la del principio:

*   **Ultra Enfocado:** De 10 columnas iniciales, nos quedamos con 5, cada una con un propósito claro para el análisis: `GRAVEDAD`, `CLASE`, `CHOQUE`, `DISENO_LUGAR` y `FECHA_HORA`.

*   **Metodológicamente Sólido:** No solo creamos una columna `datetime` funcional, sino que hemos empezado a tratar los códigos como lo que son: **categorías**. La columna `CHOQUE` ya no es un `float` ambiguo, sino una `category` descriptiva, que es el tipo de dato correcto y eficiente.

**Resumen de Habilidades de la Clase 1:**

En esta sesión, has aprendido a ser un verdadero "detective de datos". Utilizando a la IA como tu copiloto, has desarrollado un proceso de pensamiento crítico para:

1.  Formular preguntas precisas para explorar y manipular datos.
2.  Diagnosticar y solucionar problemas de tipos de datos, valores nulos y alta cardinalidad.
3.  Usar recursos externos (el `DICCIONARIO`) para descifrar y enriquecer tus datos.
4.  Transformar columnas de códigos numéricos en categorías legibles y eficientes.

---

### **Ejercicio Final (¡Tu Turno para Consolidar!)**

Antes de dar por terminada la clase, te dejamos una tarea para que practiques y consolides lo aprendido. Nuestro DataFrame está casi perfecto, pero las columnas `GRAVEDAD`, `CLASE` y `DISENO_LUGAR` todavía son de tipo `int64` (números).

**Tu misión es aplicar exactamente el mismo proceso que hicimos en el Paso 10 para transformar estas tres columnas a tipo `category` con sus descripciones correspondientes.**

Para cada columna (`GRAVEDAD`, `CLASE`, `DISENO_LUGAR`):
1.  Filtra el `df_diccionario` para obtener sus códigos y descripciones.
2.  Crea el diccionario de mapeo.
3.  Usa el método `.map()` para reemplazar los códigos en `df_siniestros`.
4.  Convierte la columna transformada al tipo `category` con `.astype('category')`.

Al terminar, tu DataFrame estará impecable y 100% listo para la siguiente etapa.

---

**En la Clase 2**, partiremos de este dataset perfectamente preparado para dejar de limpiar y empezar a descubrir: responderemos preguntas de negocio, buscaremos patrones en el tiempo y crearemos nuestras primeras visualizaciones de datos impactantes.

¡Nos vemos en la próxima clase!

In [ ]:
df_siniestros.head()

GRAVEDAD  CLASE    CHOQUE  DISENO_LUGAR          FECHA_HORA
0         2      2       NaN             2 2015-01-01 01:05:00
1         2      3       NaN             1 2015-01-01 05:50:00
2         2      3       NaN             1 2015-01-01 07:15:00
3         3      1  Vehiculo             1 2015-01-01 09:30:00
4         2      1  Vehiculo             2 2015-01-01 09:45:00